### PR TITLE
Add better logging and handle git push failures better

### LINKED ISSUE #49: Notify only if git push is successful
## Overview

If a problem occurs with pushing the local locker to the remote locker then we should not send results notifications because those notifications contain remote repository links that are broken.  They also tell a story that is incomplete because the notification information has not actually made it to the evidence locker.  Instead we should send notifications saying that the remote locker push failed and here's why...

## Requirements

- If running in `full-remote` mode send check result notifications only if the locker push was successful.
- If running in any other mode than `full-remote` send notifications as expected.
- If an unsuccessful push occurs then send notifications to all specified notifiers except for LockerNotifier detailing the problem.

## Approach

- Create a `exceptions.LockerPushError()`
- `Locker.push()`
   - Raise LockerPushError if [problem with push](https://gitpython.readthedocs.io/en/stable/reference.html#git.remote.PushInfo).

```python
    def push(self):
        """Push the local git repository to the remote repository."""
        if self._do_push:
            info = self.repo.remotes[0].push()[0]
            if info.flags >= PushInfo.ERROR:
                raise LockerPushError(info) 
```

- `CheckMode.__exit__()`
   - Catch and set `self.results` to LockerPushError
   - **~~Need to consider m~~Moving `self.fix_failures` inside the try block and _after_ `self.build_reports()`**

```python
    def __exit__(self, typ, val, traceback):
        """Handle post check test execution processing."""
        super(CheckMode, self).__exit__(typ, val, traceback)
        try:
            self.build_reports()
            self.fix_failures()
        except LockerPushError as lpe:
            # log error
            # set self.push_error = True
        self.run_notifiers()
```

- Notifiers...
   - ~~Possibility that a~~All need to be handled individually.
   - ~~Will try to see if we can handle this in `_BaseNotifier._messages_by_accreditations` - Needs more research~~
   - `FDNotifier.notify()`
      - If a locker push error encountered then notify locker push error details for each accreditation in self._results
   - `LockerNotifier.notify()`
      - Do not notify
   - `GHIssuesNotifier.notify()` 
      - If a locker push error encountered and when running in "summary by accreditation" mode then notify locker push error details for each accreditation in self._results
      - Do not notify when running notifier in "per check" mode.
   - `SlackNotifier.notify()`
      - If a locker push error encountered then notify locker push error details for each accreditation in self._results
   - `PagerDutyNotifier.notify()`
      - Do not notify
   - `FindingsNotifier.notify()`
      - Do not notify

## Security and Privacy

N/A

## Test Plan

TBD

### LINKED ISSUE #53: Provide better git logging around push failures
## Overview

When the locker push fails, it doesn't appear as though the logs show the failure so people think that the local locker was pushed to the remote locker when in fact it hasn't.

## Requirements

When a locker push fails we should raise a proper exception for that.

## Approach

TBD

## Security and Privacy

N/A

## Test Plan

TBD

### PR DESCRIPTION
**Pull request by alfinkel**
- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)

## What

Locker git operation logging was added as was logging for all of the notifiers.  Additionally, notifier behavior was modified to handle when a git remote locker push fails more gracefully.

## Why

Logging the process is always helpful for triaging execution issues as is better notifier behavior around remote locker push failures.

## How

- Add locker git ops logging
- Add notifier logging
- Ensure STDOUT notifier always runs
- Add remote locker push failure notification behavior
- Fix links in CONTRIBUTING.md
- Fix formatting and content in coding standards rst
- Update notifier unit tests

## Test

- Original execution/notification behavior is not compromised
- When a git remote locker push fails, the FD (STDOUT) notifier, Slack notifier, and GH Issues (summary by accred version) all notify by accreditation that all check results and evidence failed to be pushed to the locker.  Whereas for GH Issues (per check), PagerDuty, and Findings notifiers all do not trigger in the event of a push failure.  Logging is provided saying so.
- A push failure ERROR is logged to the execution log wit details from git.

## Context

- Closes #49 
- Closes #53

### CODE REVIEW

File: compliance/locker.py [hunk 1]
```diff
@@ -180,11 +183,10 @@ def init(self):
 
     def logger_init_msgs(self):
         """Log locker initialization information."""
-        self.logger.info(f'Local locker location is {self.local_path}')
         gpgsign = self.gitconfig.get('commit', {}).get('gpgsign')
         if self._do_push and not gpgsign:
             self.logger.warning(
-                'Commits will not be cryptographically signed, best '
+                'Commits may not be cryptographically signed, best '
```
Comment (drsm79) [inline compliance/locker.py:189]: Under what circumstances would the be signed if gpgsign is not set?

File: compliance/scripts/compliance_cli.py [hunk 2]
```diff
@@ -108,6 +108,8 @@ def _init_arguments(self):
         )
 
     def _validate_arguments(self, args):
+        if 'stdout' not in args.notify:
```
Comment (drsm79) [inline compliance/scripts/compliance_cli.py:111]: What about not having stdout in the list the user can choose from? Should update the help at line 94 to say stdout is always used.

Something for another time, if we were on py3.8 we could do something like:

```
import argparse
parser = argparse.ArgumentParser(prog='game.py')
parser.add_argument('--notify', choices=['gh_issue', 'slack', 'pagerduty'], nargs='*', action='extend')
parser.parse_args('--notify gh_issue pagerduty --notify slack'.split())
```

File: compliance/notify.py [hunk 3]
```diff
@@ -345,7 +368,7 @@ class LockerNotifier(_BaseMDNotifier):
       that they belong to.
     """
 
-    def __init__(self, results, controls, locker):
+    def __init__(self, results, controls, locker, push_error=False):
```
Comment (drsm79) [inline compliance/notify.py:371]: default to False consistently?

File: compliance/locker.py (hunk 1 above)
Comment (alfinkel) [reply compliance/locker.py:189]: When running locally you're in control of your commit signing regardless of whether you've added it to the framework config.  Adding it to the framework config ensure that the commits are signed when executed from a CI.  So the warning message is misleading if you ever run from your own environment.  That's why I figured we switch "will" to "may" as the message is misleading.  It isn't checking whether your commits will be "cryptographically signed" it is just checking whewther you've included `gpgsign` in the framework execution json.

Comment (drsm79) [reply compliance/locker.py:189]: Ah, gotcha

File: compliance/scripts/compliance_cli.py (hunk 2 above)
Comment (alfinkel) [reply compliance/scripts/compliance_cli.py:112]: I updated the help to exclude "stdout" from the valid list of values and added a note saying that the stdout notifier will always run.  There's nothing stopping someone from specifying "stdout" as a notifier which lends to backwards compatibility.  I think that's a good thing.

**Review by drsm79 — DISMISSED** [review]

**Review by drsm79 — APPROVED** [review]

**Pull request merged**